pages/_app.js

```diff
@@ -1,10 +1,48 @@
+import { useState } from "react";
+import { nanoid } from "nanoid";
+
 import GlobalStyle from "../components/GlobalStyle";
+import { shopping_list_DB } from "../services/db.js";
 
 function MyApp({ Component, pageProps }) {
+  const [shoppingListItems, setShoppingListItems] = useState(shopping_list_DB);
+
+  function toggleItemChecked(id) {
+    const updatedItems = shoppingListItems.filter((item) => item.id !== id);
+    const toggledItem = shoppingListItems.find((item) => item.id === id);
+    toggledItem.checked = !toggledItem.checked;
+    updatedItems.unshift(toggledItem);
+
+    setShoppingListItems(updatedItems);
+  }
```

Comment:
Can you find a method to toggle the checked state without removing and then adding the relevant item again? (You might have a look into the zustand demo code to find a way). We normally want to avoid these patterns, since they are not pure.

Reply:
> Can you find a method to toggle the checked state without removing and then adding the relevant item again? (You might have a look into the zustand demo code to find a way). We normally want to avoid these patterns, since they are not pure.

I used this variant, because additionally of changing the array object, I also want it to shift to the first index, so it will be displayed on top of the list, when mapped for rendering. Another solution that comes to my mind is this one:
```js
    const idx = null;
    const updatedItems = shoppingListItems.map((item, index) => {
      if (item.id === id) {
        idx = index;
        return { ...item, checked: !item.checked };
      } else {
        return { ...item };
      }
    });

    updatedItems.unshift(updatedItems.splice(idx, 1)[0]);

```

Is this a better way to do so?

Reply:
Ah well, if thats your goal your solution is fine, instead of unshift you could do: 
```
const newUpdatedItems = [
  toggledItem,
  ...updatedItems
]
setShoppingListItems(newUpdatedItems)
```